content/go/concepts/functions/functions.md

```diff
@@ -0,0 +1,339 @@
+}
+
+// passing a function as a parameter
+func printResult(f func(int64, int64) (int64, int64), a int64, b int64) {
+	c, d := f(a, b)
+	fmt.Println(c, d)
+}
+
+// functions can return functions as well!
+func returnFunc() func() {
+	return func() {
+		fmt.Println("Hey I am a function returned by function sounds cool right?")
+	}
+}
+
+func main() {
+	returnFunc()()
+	printResult(MultipleReturns, 10, 3)
+	fmt.Println(Hello("nice right"))
+	fmt.Println(MultipleReturns(10, 3))
+	fmt.Println(MultipleReturns2(10, 3))
+}
+
+```
+output is:
+
+```bash
+Hey I am a function returned by function sounds cool right?
+13 7
+check me i dont have parenthesis around my return see nice right
+13 7
+13 7
+
+Program exited.
+```
+
+Note:Functions must be directly declared at package level, not inside other functions.
+
+```go
+package main
+
+import "fmt"
+
+func main() {
+	func name() {
+		fmt.Println("hey")
+	}()
+}
+
+```
+
+```bash
+./prog.go:6:7: syntax error: unexpected name, expecting (
+./prog.go:8:3: syntax error: unexpected ( after top level declaration
+
+Go build failed.
+```
+
+we can define a function inside a function but it is not recommended as we can't call it outside the function.
+
+```go
+package main
+
+import "fmt"
+
+func main() {
+	func() {
+		fmt.Println("Hey I am no name")
+	}()
+	greeting := "hey I am no name"
+	//greeting can be accessed inside of nested functions and can be changed
+	//or assign it to a variable to make it reusable
+	secondWay := func(name string) {
+		greeting = "Hey " + name
+
+	}
+	secondWay("Pavan")
+	fmt.Println(greeting)
+
+}
+
+```
+output is 
+
+```bash
+Hey I am no name
+Hey Pavan
+
+Program exited.
+```
+
+## Anonymous Functions
+
+Anonymous functions are similar to normal functions, but do not have a name. They can be called immediately after being defined.
+Anyonymous functions can be passed as parameter to another function.
+They are also called closures/functions literals.
+
+Each function can have at most one variadic parameter(keys ...int).
+The type of a variadic parameter is always a slice type.
+think about it we have been using a function that takes vardiac parameters..
+yaa the built-in fuction Println is an example of vardiac function.
+
+Println is a variadic function refer [here](https://golang.org/pkg/fmt/#Println)
+
+for more specific information [here](https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/fmt/print.go;l=293)
+
+```go
+// You can edit this code!
+// Click here and start typing.
+package main
+
+import "fmt"
+
+func main() {
+
+	x, y := func() (int, int) {
+		fmt.Println("This function has no parameters.")
+		return 3, 4
+	}()
+	fmt.Println(x, y) //3 and 4
+
+	func() {
+		fmt.Println("x*y =", x*y)
+	}()
+
+	// The following anonymous function have no results. so it can be omitted
+
+	func(a, b int) { // equivalent to func(a int,b int)
+		println("a+b =", a+b)
+	}(x, y)
+
```

Comment:
Remove the extra line breaks. Don't have multiple line breaks in a row in the code examples.